naps/cores/cmv12k/s7_phy.py

```diff
@@ -0,0 +1,471 @@
+        self.lanes = Signal(num_lanes+1) # control lane is top lane
+
+        # signal outputs
+        self.output = [Signal(12) for _ in range(num_lanes+1)]
+        self.output_valid = Signal() # every other cycle
+
+        # control inputs
+        self.lane_pattern = Signal(12) # except for control lane
+        self.lane_delay_reset = Signal(num_lanes+1) # reset delay value to 0
+        self.lane_delay_inc = Signal(num_lanes+1) # increment delay count by 1
+        self.lane_bitslip = Signal(num_lanes+1) # slip serdes by one bit
+
+        self.outclk_delay_reset = Signal()
+        self.outclk_delay_inc = Signal()
+        self.halfswap = Signal() # swap order of half-words
+
+        # control outputs
+        self.lane_match = Signal(num_lanes+1)
+        self.lane_mismatch = Signal(num_lanes+1)
+
+    def elaborate(self, platform):
+        m = Module()
+
+        dom_in = self.domain+"_in"
+        dom_bit = self.domain+"_bit"
+        dom_hword = self.domain+"_hword"
+        dom_ctrl = self.domain+"_ctrl"
+
+        # set up the delay module for the input clock
+        m.submodules.delay_ctrl = IDelayCtrl("cmv12k_delay_ref")
+
+        delay_clk = m.submodules.delay_clk = DomainRenamer(dom_ctrl)(IDelayIncremental())
+        m.domains += ClockDomain(self.domain+"_in")
+        m.d.comb += [
+            delay_clk.reset.eq(self.outclk_delay_reset),
+            delay_clk.step.eq(self.outclk_delay_inc),
+
+            delay_clk.input.eq(self.outclk),
+            ClockSignal(self.domain+"_in").eq(delay_clk.output),
+        ]
+
+        # generate the derived clocks with a PLL
+        pll = m.submodules.pll = Mmcm(125e6, 6, 1, input_domain=self.domain+"_in")
+        pll.output_domain(dom_bit, 3)
+        pll.output_domain(dom_hword, 18)
+
+        halfswap_sync = m.submodules.halfswap_sync = PulseSynchronizer(dom_ctrl, dom_hword)
+        curr_halfswap = Signal()
+        m.d.comb += halfswap_sync.i.eq(self.halfswap)
+        m.d[dom_hword] += [
+            self.output_valid.eq(~self.output_valid),
+            curr_halfswap.eq(curr_halfswap ^ halfswap_sync.o),
+        ]
+
+        for lane in range(len(self.lanes)):
+            delay_lane = m.submodules["delay_lane_"+str(lane)] = DomainRenamer(dom_ctrl)(IDelayIncremental())
+            m.d.comb += [
+                delay_lane.reset.eq(self.lane_delay_reset[lane]),
+                delay_lane.step.eq(self.lane_delay_inc[lane]),
+
+                delay_lane.input.eq(self.lanes[lane]),
+            ]
+
+            iserdes = m.submodules["iserdes_"+str(lane)] = ISerdes(dom_bit, dom_hword)
+            bitslip_sync = m.submodules["bitslip_sync_"+str(lane)] = PulseSynchronizer(dom_ctrl, dom_hword)
+            m.d.comb += [
+                iserdes.input.eq(delay_lane.output),
+
+                bitslip_sync.i.eq(self.lane_bitslip[lane]),
+                iserdes.bitslip.eq(bitslip_sync.o),
+            ]
+
+            with m.If(self.output_valid ^ curr_halfswap):
```

Comment:
I don't think this code is correct for non periodic signals.
For the `curr_halfswap == 0` case bits `[6:12]` will contain older data than `[0:6]` and for `curr_halfswap == 1` the opposite is true. 
But I don't think the (temporal) order of the bits should ever change, instead I think `halfswap` should be more like a pulse that generate a `bitslip` by `6` bits, by simply writing the the same output halfword twice for one operation.

Am I missing something?

Reply:
fixed, possibly?